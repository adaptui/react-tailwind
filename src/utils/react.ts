import * as React from "react";

import { isFunction } from "./assertions";
import { cx } from "./tailwindMerge";
import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};
