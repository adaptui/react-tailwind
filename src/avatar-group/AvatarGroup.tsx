import { Avatar } from "../avatar";
import { forwardRefWithAs } from "../utils";

import { AvatarGroupContextProvider } from "./__utils";
import { AvatarGroupProps, useAvatarGroupProps } from "./AvatarGroupProps";
import { AvatarGroupWrapper } from "./AvatarGroupWrapper";

export const AvatarGroup = forwardRefWithAs<
  AvatarGroupProps,
  HTMLDivElement,
  "div"
>((props, ref) => {
  const { childrenWithinMax, excessChildrenCount, uiProps, wrapperProps } =
    useAvatarGroupProps(props);

  return (
    <AvatarGroupWrapper ref={ref} {...uiProps} {...wrapperProps}>
      <AvatarGroupContextProvider {...uiProps}>
        {childrenWithinMax}
        {excessChildrenCount > 0 ? (
          <Avatar
            name={excessChildrenCount.toString()}
            data-testid="testid-excess_label"
          />
        ) : null}
      </AvatarGroupContextProvider>
    </AvatarGroupWrapper>
  );
});

AvatarGroup.displayName = "AvatarGroup";
