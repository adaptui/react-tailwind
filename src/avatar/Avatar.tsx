import * as React from "react";
import { cx } from "@renderlesskit/react";

import { useTheme } from "../theme";
import { Box, BoxProps } from "../box";
import { GenericAvatar } from "../icons";
import { AvatarName } from "./AvatarName";
import { AvatarIcon } from "./AvatarIcon";
import { AvatarImage } from "./AvatarImage";
import { useImage } from "../utils/useImage";
import { useAvatarGroup } from "./AvatarGroup";
import { forwardRefWithAs } from "../utils/types";
import { createContext, runIfFn } from "../utils";
import { AvatarBadge } from "./AvatarBadge";

export type AvatarInitialProps = {
  /**
   * How large should avatar be?
   * @default "md"
   */
  size?: keyof Renderlesskit.GetThemeValue<"avatar", "size">;
  /**
   * URL for the avatar image
   */
  src?: string;
  /**
   * Name prop used for `alt` & calculate placeholder initials.
   */
  name?: string;
  /**
   * Defines loading strategy
   */
  loading?: "eager" | "lazy";
  /**
   * Function to get the initials to display
   */
  getInitials?: (name: string) => string;
  /**
   * Function called when image failed to load
   */
  onError?: () => void;
  /**
   * The default avatar used as fallback when `name`, and `src`
   * is not specified.
   * @type React.ReactElement
   */
  fallback?: React.ReactNode;
  status?: "online" | "sleep" | "typing";
  position?: "top-left" | "top-right" | "bottom-right" | "bottom-left";
};

export type AvatarProps = Omit<BoxProps, "onError"> & AvatarInitialProps;

export const Avatar = forwardRefWithAs<AvatarProps, HTMLDivElement, "div">(
  (props, ref) => {
    const {
      name,
      src,
      size,
      onError,
      fallback,
      getInitials = initials,
      loading,
      status,
      position,
      children,
      className,
      ...rest
    } = props;
    const group = useAvatarGroup();
    const _size = size || group?.size || "md";
    const theme = useTheme();
    const avatarStyles = cx(
      theme.avatar.base,
      theme.avatar.size[_size],
      className,
    );

    const context: AvatarInitialProps = React.useMemo(
      () => ({
        size: _size,
        src,
        name,
        loading,
        getInitials,
        onError,
        fallback,
        status,
        position,
      }),
      [
        _size,
        src,
        name,
        loading,
        getInitials,
        onError,
        fallback,
        status,
        position,
      ],
    );

    /**
     * Use the image hook to only show the image when it has loaded
     */
    const imageStatus = useImage({ src, onError });
    const hasLoaded = imageStatus === "loaded";

    /**
     * Fallback avatar applies under 2 conditions:
     * - If `src` was passed and the image has not loaded or failed to load
     * - If `src` wasn't passed
     *
     * In this case, we'll show either the name avatar or default avatar
     */
    const showFallback = !src || !hasLoaded;

    return (
      <AvatarProvider value={context}>
        <Box ref={ref} className={avatarStyles} {...rest}>
          {children ? (
            runIfFn(children, context)
          ) : (
            <>
              {!showFallback ? (
                <AvatarImage />
              ) : name ? (
                <AvatarName />
              ) : (
                <AvatarIcon>
                  {fallback ? fallback : <GenericAvatar />}
                </AvatarIcon>
              )}
              {status ? <AvatarBadge /> : null}
            </>
          )}
        </Box>
      </AvatarProvider>
    );
  },
);

export type AvatarContext = AvatarInitialProps;

const [AvatarProvider, useAvatarContext] = createContext<AvatarContext>({
  name: "AvatarContext",
  strict: false,
});

export { useAvatarContext };

export function initials(name: string) {
  const [firstName, lastName] = name.split(" ");
  return firstName && lastName
    ? `${firstName.charAt(0)}${lastName.charAt(0)}`
    : firstName.charAt(0);
}
