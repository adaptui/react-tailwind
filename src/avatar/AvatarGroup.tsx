import * as React from "react";
import { cx } from "@renderlesskit/react";

import { useTheme } from "../theme";
import { Box, BoxProps } from "../box";
import { forwardRefWithAs } from "../utils/types";
import { createContext, getValidChildren } from "../utils";
import { Avatar, AvatarContents, AvatarProps } from "./Avatar";

export type AvatarGroupContext = Pick<
  AvatarProps,
  "size" | "showBorder" | "borderColor"
> & {};

const [AvatarGroupProvider, useAvatarGroup] = createContext<AvatarGroupContext>(
  {
    strict: false,
    name: "AvatarGroupProvider",
  },
);

export { useAvatarGroup };

export type AvatarGroupProps = BoxProps &
  AvatarGroupContext & { limit?: number };

export const AvatarGroup = forwardRefWithAs<
  AvatarGroupProps,
  HTMLDivElement,
  "div"
>((props, ref) => {
  const {
    limit = 100,
    size = "md",
    showBorder = true,
    borderColor,
    className,
    children,
    ...rest
  } = props;
  const theme = useTheme();
  const context = React.useMemo(() => ({ size, showBorder, borderColor }), [
    size,
    showBorder,
    borderColor,
  ]);

  const validChildren = getValidChildren(children);

  /**
   * Get the avatars within the max
   */
  const childrenWithinMax = limit
    ? validChildren.slice(0, limit)
    : validChildren;

  /**
   * Get the remaining avatar count
   */
  const excess = limit != null && validChildren.length - limit;

  return (
    <AvatarGroupProvider value={context}>
      <Box
        ref={ref}
        role="group"
        aria-label="Avatar group"
        className={cx(theme.avatar.group.base, className)}
        {...rest}
      >
        {childrenWithinMax}
        <AvatarExcess
          size={size}
          excess={excess}
          {...validChildren[limit].props}
        />
      </Box>
    </AvatarGroupProvider>
  );
});

const AvatarExcess = ({
  excess,
  size = "md",
  ...props
}: AvatarProps & {
  excess: number | false;
}) => {
  const theme = useTheme();

  return excess > 0 ? (
    <Avatar {...props}>
      <>
        <AvatarContents />
        <Box className={theme.avatar.group.excess.bg} />
        <Box
          data-testid="testid-excess_label"
          className={cx(
            theme.avatar.group.excess.text.base,
            theme.avatar.group.excess.text.size[size],
          )}
        >
          +{size === "xs" ? "" : excess}
        </Box>
      </>
    </Avatar>
  ) : null;
};
