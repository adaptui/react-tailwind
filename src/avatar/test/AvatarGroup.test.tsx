import * as React from "react";
import { Avatar } from "../Avatar";
import { configureAxe } from "jest-axe";
import { render, screen } from "@testing-library/react";
import { AvatarGroup } from "../AvatarGroup";

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

describe("<AvatarGroup />", () => {
  expect.assertions(1);

  it("should render correctly", () => {
    const { asFragment } = render(
      <AvatarGroup>
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should pass the size to all avatars", () => {
    const { asFragment } = render(
      <AvatarGroup size="lg">
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should be able to overwrite specific avatar size", () => {
    const { asFragment } = render(
      <AvatarGroup size="lg">
        <Avatar />
        <Avatar size="xs" />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render truncated avatar counter if avatar count is less than limit", () => {
    render(
      <AvatarGroup limit={5}>
        <Avatar />
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-truncated");
    expect(truncateCircle).not.toBeInTheDocument();

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(3);
  });

  it("should render truncated avatar counter if avatar count is greater than limit", () => {
    render(
      <AvatarGroup limit={2}>
        <Avatar />
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-truncated");
    expect(truncateCircle).toHaveTextContent("+1");

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(3);
  });

  it("should not render truncated avatars counter if limit is equal to avatars count", () => {
    render(
      <AvatarGroup limit={2}>
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-truncated");
    expect(truncateCircle).not.toBeInTheDocument();

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(2);
  });

  it("should not have axe violations", async () => {
    const { container } = render(
      <AvatarGroup>
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
