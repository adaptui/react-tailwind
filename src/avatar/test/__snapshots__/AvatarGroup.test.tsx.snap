// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarGroup /> should be able to overwrite specific avatar size 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="flex items-center space-x-2"
    role="group"
  >
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-3xl h-20 w-20"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-sm h-8 w-8"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AvatarGroup /> should pass the size to all avatars 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="flex items-center space-x-2"
    role="group"
  >
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-3xl h-20 w-20"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-3xl h-20 w-20"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AvatarGroup /> should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="flex items-center space-x-2"
    role="group"
  >
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-2xl h-16 w-16"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
    <div
      class="font-sans font-bold inline-flex items-center relative justify-center text-center uppercase text-white bg-gray-400 fill-current rounded-full ring-2 ring-white text-2xl h-16 w-16"
    >
      <svg
        class="w-em h-em inline-block leading-em flex-shrink-0 text-current align-middle"
        focusable="false"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
          fill="currentColor"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
